File: packages/sdk-typescript/src/agent.ts
// SPDX-License-Identifier: Apache-2.0
import { existsSync } from 'node:fs'
import { join } from 'node:path'
import type { ZodTypeAny } from 'zod'
import { generateKeypair, saveKeypair, loadKeypair, encodePublicKey } from './keys.js'
import { SkillRegistry } from './skill-registry.js'
import { TaskStore } from './task-store.js'
import { RateLimiter } from './rate-limiter.js'
import { NonceStore } from './nonce-store.js'
import { buildAgentCard } from './card.js'
import { buildServer } from './server.js'
import type { CommunicationMode, TrustTier } from './types.js'
import type { FastifyInstance } from 'fastify'

export interface AgentConfig {
  name: string
  version?: string
  description?: string
  url: string
  specializations?: string[]
  models?: Array<{ provider: string; model: string }>
  keysDir?: string
  cardTTL?: number
  rateLimit?: { requestsPerMinute: number; requestsPerSender: number; tokensPerSenderPerDay?: number }
}

export interface AgentSkillOptions {
  name: string
  description: string
  input: ZodTypeAny
  output: ZodTypeAny
  modes: CommunicationMode[]
  trust: TrustTier
  allowedPeers?: string[]
  handler: (input: unknown, ctx: { sender: string; traceId: string; spanId: string; delegationToken?: string }) => Promise<unknown>
}

export interface ServeOptions {
  port?: number
  host?: string
}

export class Agent {
  private registry = new SkillRegistry()
  private config: AgentConfig
  private pendingPeers = new Map<string, Uint8Array>()

  constructor(config: AgentConfig) {
    this.config = config
  }

  skill(id: string, opts: AgentSkillOptions): this {
    this.registry.register(id, opts)
    return this
  }

  trustPeer(agentId: string, publicKey: Uint8Array): this {
    this.pendingPeers.set(agentId, publicKey)
    return this
  }

  async serve(opts: ServeOptions = {}): Promise<FastifyInstance> {
    const port = opts.port ?? 3000
    const host = opts.host ?? '0.0.0.0'
    const keysDir = this.config.keysDir ?? join(process.cwd(), '.samvad', 'keys')

    // Load or generate keypair
    let kp
    const keyFile = join(keysDir, 'key-current.json')
    if (existsSync(keyFile)) {
      kp = await loadKeypair(keysDir, 'key-current')
    } else {
      kp = await generateKeypair('key-current')
      await saveKeypair(kp, keysDir)
    }

    const card = buildAgentCard({
      name: this.config.name,
      version: this.config.version ?? '1.0.0',
      description: this.config.description ?? '',
      url: this.config.url,
      specializations: this.config.specializations ?? [],
      models: this.config.models ?? [],
      skills: this.registry.getDefs(),
      publicKeys: [{ kid: kp.kid, key: encodePublicKey(kp.publicKey), active: true }],
      rateLimit: this.config.rateLimit ?? { requestsPerMinute: 60, requestsPerSender: 10 },
      cardTTL: this.config.cardTTL ?? 300,
    })

    const introText = `# ${card.name}\n\n${card.description}\n\n## Skills\n${card.skills.map(s => `- **${s.name}**: ${s.description}`).join('\n')}`

    // Merge the agent's own public key (for self-calls in tests) with trusted peers
    const knownPeers = new Map(this.pendingPeers)
    knownPeers.set(card.id, kp.publicKey)

    const server = buildServer({
      card,
      registry: this.registry,
      keypair: kp,
      taskStore: new TaskStore(3600_000),
      rateLimiter: new RateLimiter(card.rateLimit),
      nonceStore: new NonceStore(5 * 60_000),
      introText,
      knownPeers,
    })

    await server.listen({ port, host })
    console.log(`[SAMVAD] Agent "${card.name}" listening on ${host}:${port}`)
    console.log(`[SAMVAD] Card: ${card.url}/.well-known/agent.json`)
    return server
  }
}

File: packages/sdk-python/src/samvad/agent.py
# SPDX-License-Identifier: Apache-2.0
from __future__ import annotations

from collections.abc import Awaitable, Callable
from pathlib import Path
from typing import Any

from pydantic import BaseModel

from .card import build_agent_card
from .keys import Keypair, load_or_generate_keypair
from .nonce_store import InMemoryNonceStore, NonceStore
from .rate_limiter import RateLimiter
from .server import ServerConfig, build_app
from .skill_registry import SkillRegistry
from .types import (
    AgentCard,
    CommunicationMode,
    InjectionClassifier,
    PublicKey,
    RateLimit,
    SkillContext,
    TrustTier,
)


class Agent:
    def __init__(
        self,
        *,
        name: str,
        description: str = "",
        url: str,
        version: str = "1.0.0",
        specializations: list[str] | None = None,
        models: list[dict[str, Any]] | None = None,
        keys_dir: str = ".samvad/keys/",
        rate_limit: RateLimit | None = None,
        injection_classifier: InjectionClassifier | None = None,
        card_ttl: int = 300,
        nonce_store: NonceStore | None = None,
    ) -> None:
        self._name = name
        self._description = description
        self._url = url
        self._version = version
        self._specializations = specializations or []
        self._models = models or []
        self._keys_dir = Path(keys_dir)
        self._rate_limit = rate_limit or RateLimit(requestsPerMinute=60, requestsPerSender=20)
        self._injection_classifier = injection_classifier
        self._card_ttl = card_ttl
        self._nonce_store = nonce_store
        self._registry = SkillRegistry()
        self._known_peers: dict[str, str] = {}  # agent:// URI → public key b64
        self._keypair: Keypair | None = None

    def skill(
        self,
        *,
        name: str,
        description: str,
        input_schema: type[BaseModel],
        output_schema: type[BaseModel],
        modes: list[CommunicationMode],
        trust: TrustTier,
        handler: Callable[[BaseModel, SkillContext], Awaitable[BaseModel]],
        allowed_peers: list[str] | None = None,
    ) -> Agent:
        self._registry.register(
            name=name,
            description=description,
            input_schema=input_schema,
            output_schema=output_schema,
            modes=modes,
            trust=trust,
            handler=handler,
            allowed_peers=allowed_peers,
        )
        return self

    def trust_peer(self, agent_id: str, public_key_b64: str) -> Agent:
        self._known_peers[agent_id] = public_key_b64
        return self

    def _ensure_keypair(self) -> Keypair:
        if self._keypair is None:
            self._keypair = load_or_generate_keypair(self._keys_dir, "agent")
        return self._keypair

    def build_card(self) -> AgentCard:
        kp = self._ensure_keypair()
        pub_key = PublicKey(kid=kp.kid, key=kp.public_key_b64, active=True)
        return build_agent_card(
            name=self._name,
            version=self._version,
            description=self._description,
            url=self._url,
            specializations=self._specializations,
            models=self._models,
            skills=self._registry.to_skill_defs(),
            public_keys=[pub_key],
            rate_limit=self._rate_limit,
            card_ttl=self._card_ttl,
        )

    def build_app(self) -> Any:  # returns Starlette
        from .task_store import TaskStore
        kp = self._ensure_keypair()
        card = self.build_card()
        config = ServerConfig(
            card=card,
            registry=self._registry,
            known_peers=self._known_peers,
            nonce_store=self._nonce_store or InMemoryNonceStore(),
            rate_limiter=RateLimiter(
                requests_per_minute=self._rate_limit.requests_per_minute,
                requests_per_sender=self._rate_limit.requests_per_sender,
                tokens_per_sender_per_day=self._rate_limit.tokens_per_sender_per_day,
            ),
            task_store=TaskStore(),
            sign_keypair=kp,
            injection_classifier=self._injection_classifier,
        )
        return build_app(config)

    async def serve(self, *, host: str = "0.0.0.0", port: int = 3002) -> None:
        import uvicorn
        app = self.build_app()
        server = uvicorn.Server(uvicorn.Config(app, host=host, port=port, log_level="info"))
        await server.serve()
